### DISCUSSION TITLE
Error Authentication Read Mifare Classic

### DISCUSSION THREAD

**Discussion by gitexpled** [capawesome-team/capacitor-plugins#303] [Q&A]
Hello,

I am currently working on an Ionic 8 project where I am attempting to authenticate with a MIFARE Classic NFC card. However, I am encountering an authentication error when trying to assign the default key provided on the NFC card. At this point, my objective is to read block 4 of the card, which could be either type A or type B. Despite recognizing the NFC device correctly through the console, the authentication process fails.

Could you please assist me in understanding the possible reasons for this error and how I can resolve the issue? Any guidance or suggestions would be greatly appreciated.

Thank you in advance for your support.

![image (1)](https://github.com/user-attachments/assets/b5ef7f0a-81c2-4ffb-b993-acccc448ff29)

this is the service in ionic:
## Service TypeScript

```typescript
import { Injectable } from '@angular/core';
import { Nfc, NfcTagTechType } from '@capawesome-team/capacitor-nfc';
import { Subject } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class NfcService {
  public nfcContentChanged = new Subject<string>();

  constructor() {
    this.startNfcListener();
  }

  private async startNfcListener(): Promise<void> {
    await Nfc.removeAllListeners();

    Nfc.addListener('nfcTagScanned', async (event: any) => {
      try {
        if (event.nfcTag.id) {
          const idArray = event.nfcTag.id;
          const nfcId = idArray
            .map((byte: number) => byte.toString(16).padStart(2, '0'))
            .join('')
            .toUpperCase();
          console.log('ID de la tarjeta NFC:', nfcId);
          await Nfc.connect({
            techType: NfcTagTechType.MifareClassic,
          });

          const blockToAuthenticate = 4;
          const defaultKey = [0xff, 0xff, 0xff, 0xff, 0xff, 0xff];
          const uid = event.nfcTag.id;

          console.log(`Autenticando bloque ${blockToAuthenticate}...`);

          try {
            await this.authenticateBlock(blockToAuthenticate, defaultKey, 'B', uid);
            console.log('Autenticación exitosa.');
          } catch (authError) {
            console.error('Error durante la autenticación:', authError);
            return;
          }
          const startBlock = 4;
          const numBlocks = 4;

          for (let i = startBlock; i < startBlock + numBlocks; i++) {
            try {
              const blockData = await this.readBlock(i);
              console.log(`Datos del bloque ${i}:`, blockData);
            } catch (readError) {
              console.error(`Error al leer el bloque ${i}:`, readError);
            }
          }
        }
      } catch (error) {
        console.error('Error al leer la tarjeta NFC:', error);
      } finally {
        await Nfc.close();
      }
    });

    await Nfc.startScanSession({
      techTypes: [NfcTagTechType.MifareClassic],
    });
  }

  private async authenticateBlock(
    blockNumber: number,
    key: number[],
    keyType: 'A' | 'B',
    uid: number[]
  ): Promise<void> {
    const command = keyType === 'A' ? 0x60 : 0x61;
    const uidReversed = uid.slice().reverse();
    const authCommand = [command, blockNumber, ...key, ...uidReversed];

    await Nfc.transceive({
      techType: NfcTagTechType.MifareClassic,
      data: authCommand,
    });
  }

  private async readBlock(blockNumber: number): Promise<string> {
    try {
      const { response } = await Nfc.transceive({
        techType: NfcTagTechType.MifareClassic,
        data: [0x30, blockNumber],
      });

      return this.bytesToString(response);
    } catch (error) {
      throw error;
    }
  }

  private bytesToString(bytes: number[]): string {
    const decoder = new TextDecoder();
    return decoder.decode(new Uint8Array(bytes));
  }

  public async checkNfcSupport(): Promise<boolean> {
    const { isSupported } = await Nfc.isSupported();
    return isSupported;
  }

  public async checkNfcEnabled(): Promise<boolean> {
    const { isEnabled } = await Nfc.isEnabled();
    return isEnabled;
  }

  public async requestPermissions(): Promise<void> {
    await Nfc.requestPermissions();
  }

  public async openSettings(): Promise<void> {
    await Nfc.openSettings();
  }
}

**Comment by robingenz** [comment]
What is the error message?

**Comment by gitexpled** [comment]
I'm working on an Ionic project that requires reading data from Mifare Classic NFC cards (both Type A and B). In a Node.js environment, I successfully authenticated and read the card data using a default key ('FFFFFFFFFFFF') and the following code:

```js
app.get('/api/nfc-read', async (req, res) => {
    if (!currentCard) {
        return res.status(400).json({ message: 'No card detected', data: null });
    }

    try {
        const { reader, card } = currentCard;
        await reader.authenticate(4, 0x60, 'FFFFFFFFFFFF');
        const data = await reader.read(4, 16, 16);

        res.json({ message: 'Read successful', data: data.toString('utf8').trim() });
    } catch (error) {
        console.error('Error reading the card:', error);
        res.status(500).json({ message: 'Error reading the card', data: null });
    }
});
```

I am now attempting to implement similar functionality in Ionic. I need to authenticate a block using either key type 'A' or 'B' and read the data from the block. My current approach is as follows:

```js
private async authenticateBlock(
  blockNumber: number,
  key: number[],
  keyType: 'A' | 'B',
  uid: number[]
): Promise<void> {
  const command = keyType === 'A' ? 0x60 : 0x61;
  const uidReversed = uid.slice().reverse();
  const authCommand = [command, blockNumber, ...key, ...uidReversed];

  await Nfc.transceive({
    techType: NfcTagTechType.MifareClassic,
    data: authCommand,
  });
}

private async readBlock(blockNumber: number): Promise<string> {
  try {
    const { response } = await Nfc.transceive({
      techType: NfcTagTechType.MifareClassic,
      data: [0x30, blockNumber],
    });

    return this.bytesToString(response);
  } catch (error) {
    throw error;
  }
}
```
I would like some guidance on properly implementing the authentication in Ionic, especially for both key types ('A' and 'B'). The Node.js approach worked with the default key, but I'm facing challenges in adapting it to work with Ionic's NFC plugin. 

this es error from ionic
![image](https://github.com/user-attachments/assets/80fc3971-6337-4b1f-a102-a459a1e81135)

I tried another configuration to read another type of card with MIFARE Ultralight and the reading worked without problems, I don't think it's a permissions issue.

**Reply by robingenz** [reply]
> I would like some guidance on properly implementing the authentication in Ionic, especially for both key types ('A' and 'B'). The Node.js approach worked with the default key, but I'm facing challenges in adapting it to work with Ionic's NFC plugin.

Unfortunately, I am not familiar with the authentication process of MIFARE Classic NFC cards but I will do my best to help you.

A few question:

- What type is the variable `reader` in your example?
- Does the error only occur under Android or also under iOS?
- What error message is displayed in Android Studio Logcat?

Here are a few ideas:

- Maybe you are doing something wrong when converting from string/hex to bytes. Make sure to use the [`NfcUtils`](https://github.com/capawesome-team/capacitor-plugins/blob/main/packages/nfc/docs/utils/README.md) class for such tasks.
- Have you asked ChatGPT? I often use it when I have problems with NFC and it works very well.
- Here is an example of how I use the plugin in another project: https://github.com/DHBW-VS/app/blob/42b82c1b7360fa62ac93696d2775e4176f8edff8/src/app/modules/canteen/components/canteen-card-balance-modal/canteen-card-balance-modal.component.ts#L62:L88 Maybe this will help you.
